pipelines/google_sheets/data_validator.py

```diff
@@ -0,0 +1,177 @@
+    for val_dict in value_list:
+        # TODO: try except
+        is_date = "effectiveValue" in val_dict and \
+                  "numberValue" in val_dict['effectiveValue'] and \
+                  "effectiveFormat" in val_dict and \
+                  "numberFormat" in val_dict['effectiveFormat'] and\
+                  "type" in val_dict["effectiveFormat"]["numberFormat"] and\
+                  ("DATE" in val_dict["effectiveFormat"]["numberFormat"]["type"] or "TIME" in val_dict["effectiveFormat"]["numberFormat"]["type"])
+        value_type_list.append(is_date)
+    return value_type_list
+
+
+def process_url(url: str) -> str:
+    """"
+    This function takes an url to a Google spreadsheet and computes the spreadsheet id from it according to the spreadsheet url formula: https://docs.google.com/spreadsheets/d/<spreadsheet_id>/edit
+    If the url is not formatted correctly a Value Error will be returned
+    @:param: url- the string containing the url to the spreadsheet
+    @:return: spreadsheet_id as a string or ValueError if the url is not properly formatted
+    """
+
+    # split on the '/'
+    parts = url.split("/")
+
+    # loop through parts
+    for i in range(len(parts)):
+        # if we find
+        if parts[i] == URL_ID_IDENTIFIER and i+1 < len(parts):
+            # if the id part is left empty then the url is not formatted correctly
+            if parts[i+1] == "":
+                raise ValueError("Spreadsheet ID is an empty string")
+            else:
+                return parts[i+1]
+    # if url cannot be found, raise error
+    raise ValueError("Invalid URL. Cannot find spreadsheet ID")
+
+
+def get_spreadsheet_id(url_or_id: str) -> str:
+    """"
+    This function receives an id or url to a Google Spreadsheet and returns the spreadsheet_id as a string
+    @:param: url_or_id a string which is the id or url of the spreadsheet
+    @:return: spreadsheet_id a string which is definetly the id of the spreadsheet
+    """
+
+    # TODO: raise value error for empty id ?
+
+    # check if this is an url: http or https in it
+    if match(r"http://|https://", url_or_id):
+        # process url
+        spreadsheet_id = process_url(url_or_id)
+        return spreadsheet_id
+    else:
+        # just return id
+        return url_or_id
+
+
+def serial_date_to_datetime(serial_number: Union[int, float]) -> datetime:
+    """
+    This function receives a serial number which can be an int or float(depending on the serial number) and outputs a datetime object
+    @:param: serial_number- int/float. The integer part shows the number of days since December 30th 1899, the decimal part shows the fraction of the day
+    @:return: converted_date: datetime object for the same date as the serial number
+    """
+    # TODO: add timezone to data
+
+    # if called with a different data type, return with whatever input was, handled by the dlt pipeline later
+    if not isinstance(serial_number, (int, float)):
```

Comment:
The first row of data will deduce which column has data types or not. Every field which is expected to be a data type will be directed to this function. If some of the data in a column is not a date despite it being supposed to be a data, it will cause the program to crash. The most common of this errors would be for the user to input a date value only for google_sheets not to recognise it as a date value and thus saving it as a string. I believe it would lead to better user experience to 'tolerate' these mistakes  and return them as the original data type since dlt will handle this anyway.

Reply:
right!